### PR TITLE
KAFKA-17750: Extend kafka-consumer-groups command line tool to support new consumer group (part 1)

### PR DESCRIPTION
**Pull request by FrankYang0529**
* Bump `validVersions` of `ConsumerGroupDescribeRequest.json` and `ConsumerGroupDescribeResponse.json` to `"0-1"`.
* Add `MemberType` field to `ConsumerGroupDescribeResponse.json`. Default value is -1 (unknown). 0 is for classic member and 1 is for consumer member.
* When `ConsumerGroupMember#useClassicProtocol` is true, return `MemberType` field as `0`. Otherwise, return `1`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/resources/common/message/ConsumerGroupDescribeRequest.json [hunk 1]
```diff
@@ -18,7 +18,7 @@
   "type": "request",
   "listeners": ["broker"],
   "name": "ConsumerGroupDescribeRequest",
-  "validVersions": "0",
+  "validVersions": "0-1",
```
Comment (dajac) [inline clients/src/main/resources/common/message/ConsumerGroupDescribeRequest.json:23]: Could you please add a note about version 1 here? Take a look at other requests/responses to see how we usually do it.

**Review by dajac — COMMENTED** [review]
@FrankYang0529 Thanks for the patch. Could you also please look into whether we can extend ConsumerGroupDescribeRequestTest to validate this change?

File: clients/src/main/resources/common/message/ConsumerGroupDescribeRequest.json (hunk 1 above)
Comment (FrankYang0529) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeRequest.json:23]: Updated the PR. Thanks for the reminder.

**Comment by dajac** [comment]
@FrankYang0529 Could you please extend tests in `ConsumerGroupDescribeRequestTest` to cover the new version of the API?

**Comment by FrankYang0529** [comment]
> Could you please extend tests in `ConsumerGroupDescribeRequestTest` to cover the new version of the API?

Sorry, I missed this comment yesterday. I update the PR and CI result looks good. Could you review when you have time? Thanks.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 2]
```diff
@@ -213,4 +216,111 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+
+    // The joining request with a consumer group member 2 is accepted.
+    consumerGroupHeartbeat(
+      groupId = groupId,
+      memberId = "member-2",
+      rebalanceTimeoutMs = 5 * 60 * 1000,
+      subscribedTopicNames = List(topicName),
+      topicPartitions = List.empty,
+      expectedError = Errors.NONE
+    ).memberId()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: It seems that we could remove `memberId()`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 3]
```diff
@@ -213,4 +216,111 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+    val actual = consumerGroupDescribe(
+      groupIds = List(groupId),
+      includeAuthorizedOperations = true,
+      version = 0.toShort,
+    )
+    assertEquals(1, actual.size)
+    val group = actual.head
+    val member1 = group.members().asScala.find(_.memberId() == memberId1)
+    assertNotNull(member1)
+    assertFalse(member1.get.isClassic)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: We usually validate the full response like we did in the other tests. Would it be possible?

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 4]
```diff
@@ -213,4 +216,111 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+        groupIds = List(groupId),
+        includeAuthorizedOperations = true,
+        version = version.toShort,
+      )
+      assertEquals(1, actual.size)
+      val group = actual.head
+      val member1 = group.members().asScala.find(_.memberId() == memberId1)
+      assertNotNull(member1)
+      assertTrue(member1.get.isClassic)
+    }
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: I wonder if we could merge this with the previous call for version 0. Then we could use define the expected value for isClassic based on the version. What do you think?

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 5]
```diff
@@ -213,4 +216,111 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+    consumerGroupHeartbeat(
+      groupId = groupId,
+      memberId = memberId1,
+      rebalanceTimeoutMs = 5 * 60 * 1000,
+      subscribedTopicNames = List("foo"),
+      topicPartitions = List.empty,
+      expectedError = Errors.NONE
+    )
+
+    // there is no classic member in the group
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: I have noticed that the style of your comments are a bit inconsistent. It would be great if they could all start with a capital letter and finish with a dot.

**Review by dajac — COMMENTED** [review]
@FrankYang0529 Thanks for the update. I left a few more minor comments for consideration.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala (hunk 2 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: I also add `isClassic` test for member 2. We need to get the `memberId`, so I leave it here. Thanks.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala (hunk 3 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: I don't compare full response for two reasons:

* It looks like we don't guarantee member order if there are multiple members. If we do assertion for group, it may fail cause of different member order.
https://github.com/apache/kafka/blob/9d23f89e05782875f5f37f1e9687c621b0e3edb8/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java#L954-L961

* For uniform assignor, the target assignment result may change.

So I think we can focus on `isClassic` field.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala (hunk 4 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: Yes, updated it. Thanks for the suggestion.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala (hunk 5 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: Updated it. Thanks for the reminder.

File: clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json [hunk 6]
```diff
@@ -69,7 +70,9 @@
             { "name": "Assignment", "type": "Assignment", "versions": "0+",
               "about": "The current assignment." },
             { "name": "TargetAssignment", "type": "Assignment", "versions": "0+",
-              "about": "The target assignment." }
+              "about": "The target assignment." },
+            { "name": "IsClassic", "type": "bool", "versions": "1+", "ignorable": true,
+              "about": "True for classic member." }
```
Comment (dajac) [inline clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: I think that there is one case which is not clear in the KIP. What value will the admin client return when the version 1 is not available? If the new version is not available, we don't know whether the member is classic or not. Hence, we cannot use true or false by default. What's your take on this?

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 7]
```diff
@@ -213,4 +216,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+
+    // The joining request with a consumer group member 2 is accepted.
+    val memberId2 = consumerGroupHeartbeat(
+      groupId = groupId,
+      memberId = "member-2",
+      rebalanceTimeoutMs = 5 * 60 * 1000,
+      subscribedTopicNames = List(topicName),
+      topicPartitions = List.empty,
+      expectedError = Errors.NONE
+    ).memberId()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: You can omit the `()`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 8]
```diff
@@ -213,4 +216,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+
+    for (version <- ApiKeys.CONSUMER_GROUP_DESCRIBE.oldestVersion() to ApiKeys.CONSUMER_GROUP_DESCRIBE.latestVersion(isUnstableApiEnabled)) {
+      val actual = consumerGroupDescribe(
+        groupIds = List(groupId),
+        includeAuthorizedOperations = true,
+        version = version.toShort,
+      )
+      assertEquals(1, actual.size)
+      val group = actual.head
+      val member1 = group.members().asScala.find(_.memberId() == memberId1)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: You can omit the `()` after `members`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 9]
```diff
@@ -213,4 +216,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+      assertFalse(member1.isEmpty)
+      // Version 0 doesn't have isClassic field, so isClassic field on member 1 is false.
+      // After version 1, there is isClassic field and it should be true for member 1.
+      assertEquals(version > 0.toShort, member1.get.isClassic)
+
+      val member2 = group.members().asScala.find(_.memberId() == memberId2)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: ditto.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 10]
```diff
@@ -213,4 +216,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+    // There is no classic member in the group.
+    for (version <- ApiKeys.CONSUMER_GROUP_DESCRIBE.oldestVersion() to ApiKeys.CONSUMER_GROUP_DESCRIBE.latestVersion(isUnstableApiEnabled)) {
+      val actual = consumerGroupDescribe(
+        groupIds = List(groupId),
+        includeAuthorizedOperations = true,
+        version = version.toShort,
+      )
+      assertEquals(1, actual.size)
+      val group = actual.head
+      val member1 = group.members().asScala.find(_.memberId() == memberId1)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: ditto.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 11]
```diff
@@ -213,4 +216,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+      assertFalse(member1.isEmpty)
+      assertFalse(member1.get.isClassic)
+
+      val member2 = group.members().asScala.find(_.memberId() == memberId2)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: ditto.

**Review by dajac — COMMENTED** [review]
@FrankYang0529 Thanks for the update. I left a few nits and one question.

File: clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json (hunk 6 above)
Comment (chia7712) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: In the RPC layer, everything should function correctly since the new broker does not return the `IsClassic` field for version 0.

The primary issue lies in how the Java client distinguishes between "false" and "unknown" values, as it always defaults to "false." To address this, we could add a `ConsumerGroupDescribeResponse#version` method that returns the response version. This would allow the caller to differentiate between "false" and "unknown" based on the version.

Additionally, the `MemberDescription#isClassic` field also needs to differentiate between "false" and "unknown." One possible solution is to use `Optional<Boolean>` instead.

Comment (FrankYang0529) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Hi @dajac / @chia7712, thanks for the review. It looks like we can implement it in Admin side. If `ConsumerGroupDescribeResponse#version` is `0`, set `MemberDescription#isClassic` as `Optional.empty()`. If `ConsumerGroupDescribeResponse#version` is `1`, set `MemberDescription#isClassic` as `Optional.of(value)`. I change the KIP to align this discussion.

Comment (dajac) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Using an Optional makes total sense. For the RPC, I was wondering whether we want to make it more explicit in the schema. For instance, we could use a int8 with three values: -1, 0, 1. -1 would be used by default and would mean no provided. Relying on the version also works but it may be easy to miss for people implementing other clients.

Comment (chia7712) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: > For instance, we could use a int8 with three values: -1, 0, 1. -1 would be used by default and would mean no provided. 

While this approach benefits the Java client, it introduces some drawbacks for the Java server and RPC layer:

1. Server Implementation Complexity: The server must pass the correct integer value to `setIsClassic(int)`, which can be easily overlooked by developers implementing the server.

2. Documentation Requirements: We need to update the documentation to specify that only 0 and 1 are valid values in RPC version 1 and above. In contrast, using a boolean with version control would be more readable and straightforward for this case.

Comment (FrankYang0529) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Hi @dajac / @chia7712, I updated the PR to use `int8`. Although this approach is more complicated, it provides more information than boolean. For both types, we need to update document to clarify that `isClassic` only works for RPC version 1 and above. I think we can focus on whether server complexity is acceptable or not. For me, it's ok.

Comment (dajac) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: I personally prefer to have a bit more complexity on the server rather than relying on all the developers of clients to do the right thing.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 12]
```diff
@@ -213,4 +219,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+
+    for (version <- ApiKeys.CONSUMER_GROUP_DESCRIBE.oldestVersion() to ApiKeys.CONSUMER_GROUP_DESCRIBE.latestVersion(isUnstableApiEnabled)) {
+      val actual = consumerGroupDescribe(
+        groupIds = List(groupId),
+        includeAuthorizedOperations = true,
+        version = version.toShort,
+      )
+      assertEquals(1, actual.size)
+      val group = actual.head
+      val member1 = group.members().asScala.find(_.memberId == memberId1)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: You can omit the `()` after `members`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 13]
```diff
@@ -213,4 +219,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+      assertFalse(member1.isEmpty)
+      // Version 0 doesn't have isClassic field, so isClassic field on member 1 is false.
+      // After version 1, there is isClassic field and it should be true for member 1.
+      assertEquals(if (version == 0) -1.toByte else 1.toByte, member1.get.isClassic)
+
+      val member2 = group.members().asScala.find(_.memberId == memberId2)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: You can omit the `()` after `members`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 14]
```diff
@@ -213,4 +219,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+    // There is no classic member in the group.
+    for (version <- ApiKeys.CONSUMER_GROUP_DESCRIBE.oldestVersion() to ApiKeys.CONSUMER_GROUP_DESCRIBE.latestVersion(isUnstableApiEnabled)) {
+      val actual = consumerGroupDescribe(
+        groupIds = List(groupId),
+        includeAuthorizedOperations = true,
+        version = version.toShort,
+      )
+      assertEquals(1, actual.size)
+      val group = actual.head
+      val member1 = group.members().asScala.find(_.memberId == memberId1)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: You can omit the `()` after `members`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala [hunk 15]
```diff
@@ -213,4 +219,108 @@ class ConsumerGroupDescribeRequestTest(cluster: ClusterInstance) extends GroupCo
...
+      assertFalse(member1.isEmpty)
+      assertEquals(if (version == 0) -1.toByte else 0.toByte, member1.get.isClassic)
+
+      val member2 = group.members().asScala.find(_.memberId == memberId2)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestTest.scala]: nit: You can omit the `()` after `members`.

**Review by dajac — APPROVED** [review]
@FrankYang0529 Thanks. Overall, LGTM. I left a few nits. @chia7712 Are you OK with the change that we discussed? If so, @FrankYang0529 could you please update the KIP and notify the thread about the change?

File: clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json (hunk 6 above)
Comment (chia7712) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: > I personally prefer to have a bit more complexity on the server rather than relying on all the developers of clients to do the right thing.

I tried to imagine the code for both schemas.

### int8

client code
```java
switch (data.isClassic()) {
  case 0: return Optional.of(false);
  case 1: return Optional.of(true);
  default: return Optional.empty();
}
```
server code
```java
data.setIsClassic(useClassicProtocol ? 1 : 0);
```
---
### bool

client code
```java
if (version < 1) return Optional.empty();
return Optional.of(data.isClassic());
```

server
```java
data.setIsClassic(useClassicProtocol);
```
---

if I'm imagining correctly, it seems that using a bool could simplify the code.

Comment (dajac) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: @chia7712 It is not about simplifying the code. It is about ensuring that we have a well defined protocol and that folks implementing the protocol on the client side do the right thing. With the byte, is is explicit in the protocol that the value maybe undefined. With the boolean, they have to think about taking into account the version in that decision. It is really easy to miss it in my opinion. We use sentinels as default values in many places in the protocol for this purpose.

Comment (chia7712) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Given that the response data may be propagated without version information, using the data itself to distinguish "undefined" values is safer. Additionally, I support using `int8`.

File: clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json [hunk 16]
```diff
@@ -69,7 +70,9 @@
             { "name": "Assignment", "type": "Assignment", "versions": "0+",
               "about": "The current assignment." },
             { "name": "TargetAssignment", "type": "Assignment", "versions": "0+",
-              "about": "The target assignment." }
+              "about": "The target assignment." },
+            { "name": "IsClassic", "type": "int8", "versions": "1+", "default": "-1", "ignorable": true,
```
Comment (chia7712) [inline clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Should we consider renaming it back to "memberType" after the type is changed to `int8`?

**Comment by FrankYang0529** [comment]
> could you please update the KIP and notify the thread about the change?

Hi @dajac, I just sent a mail to vote thread and updated KIP. Thanks.

File: clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json (hunk 16 above)
Comment (FrankYang0529) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Hi @chia7712, I prefer to use `IsClassic` because it's more Intuitive. Also, we don't plan to expand the field as `protocol`, so I think we can keep it as `IsClassic`.

Comment (chia7712) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: >  I prefer to use IsClassic because it's more Intuitive.

I agree that using `IsClassic=false` and `IsClassic=true` would be more intuitive. However, currently, `IsClassic` can take the values `-1`, `0`, and `1`. These "magic numbers" are unclear without referring to the documentation.

Additionally, the codebase uses these magic numbers inconsistently across different classes. For example, several classes such as `ScramMechanism`, `EndpointType`, `ResourceType`, and `AclOperation` use `0` to represent "unknown" or "none". `MemberState` use `127` for unknown. 

In summary, it seems to me the name `IsClassic` is no longer intuitive after adopting the int8 type.

Comment (FrankYang0529) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Hi @chia7712, thanks for the suggestion. Yes, it's good to follow naming convention. I update both PR and KIP.

Comment (dajac) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: I agree with @chia7712. Small nit. I wonder if we should now use 0 for classic member and 1 for consumer member as classic is the older one.

Comment (chia7712) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: >  I wonder if we should now use 0 for classic member and 1 for consumer member as classic is the older one.

+1, that order is intuitive to me

Comment (FrankYang0529) [reply clients/src/main/resources/common/message/ConsumerGroupDescribeResponse.json]: Thanks for the suggestion. Updated both PR and KIP.

**Review by dajac — APPROVED** [review]
LGTM!

@FrankYang0529 Could you please update the description?

**Comment by FrankYang0529** [comment]
Yes, updated it. Thanks.

**Review by chia7712 — APPROVED** [review]
